### ISSUE TITLE
It should be possible to opt-out lazy-loading per client, per resource and/or per query

### ISSUE THREAD

**Issue by timorzadir** [okb/Pomona#56]

**Comment by BeeWarloc** [comment]
Jeg lurer på om det heller bør ligge i en .Settings property, for å ikke forkludre IPomonaClient interfacet for mye?

```
restClient.Settings.LazyLoadingEnabled = false; //Disable lazy for entire client
```

**Comment by jarlrasm** [comment]
A couple of thoughts:

I don't like EnableLazyLoading(false) from the timorzadirs suggestion.Should be DisableLazyLoading()
I don't think having the possibility to enable/disable for a single resource-type is useful beyond creating clutter.
A .Settings or .Options property would be much cleaner, yes please.

**Comment by timorzadir** [comment]
Agree on the .Settings and the DisableLazyLoading(). 

The enable/disable for a single-resource was simply to make the hierarchy of options complete.  
Per client and per query options will provide enough flexibilit, and both are highly needed.

**Comment by BeeWarloc** [comment]
Sorry. This was closed a bit prematurely.
As of f8dec92493a8fad16eafd9bc2eb4f809a36975b4 the basic support for turning on or off lazy loading support in client should be supported.

**Comment by BeeWarloc** [comment]
@asbjornu could you add some comments on what is left for full feature completion?

**Comment by asbjornu** [comment]
Given @timorzadir's example, the thing that is completed is:

```csharp
var restClient = new RestClient.Client("http://example.com", new HttpWebRequestClient());
restClient.LazyLoadingEnabled = false; //Disable lazy for entire client
```
I.e. "opt-out lazy-loading per client". The syntax is a bit different (`restClient.Settings.LazyMode = LazyMode.Disabled`), but the above is now supported. What's missing is more fine-grained lazy support, i.e. "opt-out lazy-loading per resource and/or per query", like:

```csharp
restClient.Customers.LazyLoadingEnabled = true; //Enable lazy for resource. Nullable. Inherits client setting if null.
var customers = restClient.Query<ICustomer>().EnableLazyLoading().ToList(); //Enable lazy for current query
var products = restClient.Query<IProduct>().EnableLazyLoading(false).ToList(); //Disable lazy for current query
```

I believe the foundation to support this is in place, but extension methods that manipulate the lazy mode state for each query need to be created along with tests to verify the behavior.

**Comment by GeirGrusom** [comment]
`restClient.Query<IProduct>(QueryFlags.Lazy)`

That way there is no need for additional extension functions and the complexities of manipulating private state of another class and the functionality is immediately clear to the API consumer. It also works better with unit-tests as it is simpler to test that `Query(Arguments)` was invoked rather than `Query().ExtensionFunction()` if lazy was enabled as part of a bug fix.

**Comment by BeeWarloc** [comment]
Yup, I agree with @GeirGrusom. It might be preferrable to make it a part of a RequestOptions parameter though, to avoid too many overloads.

**Comment by asbjornu** [comment]
I agree that it belongs in `RequestOptions` as the place in which the request itself is modified, but I am not sure how the `RequestOptions` object should be manipulated.
